### PR TITLE
Feature/nyan 85 친구관계 관련 api 기획 후 생성

### PR DESCRIPTION
**Pull request by dicohy27**

### CODE REVIEW

**Comment by uuccu** [comment]
### Code Coverage
|File|Coverage [0.00%]|
|:-|:-:|
|[src/main/kotlin/fittibackendapp/domain/follow/entity/Follow.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fentity%2FFollow.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/graphql/mutation/FollowMutationMapper.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fgraphql%2Fmutation%2FFollowMutationMapper.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/graphql/query/FollowQueryMapper.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fgraphql%2Fquery%2FFollowQueryMapper.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fservice%2FFollowService.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/dto/FollowDto.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdto%2FFollowDto.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/exception/AlreadyExistingFollowException.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fexception%2FAlreadyExistingFollowException.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/exception/NotFoundFollowException.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/37c0e9fb5cb016e0a6b0378f8fe704151c268fe8/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fexception%2FNotFoundFollowException.kt)|0.00%|

|Total Project Coverage|0.00%|:x:|
|:-|:-:|:-:|

File: src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        followerId: Long,
+        followeeId: Long,
+    ): FollowDto {
+        val existingFollow = followRepository.findByFollowerIdAndFolloweeId(followerId, followeeId)
+        if (existingFollow != null) {
+            throw AlreadyExistingFollowException()
+        }
+
+        val follower = userRepository.findByIdOrNull(followerId) ?: throw NotFoundUserException()
+        val followee = userRepository.findByIdOrNull(followeeId) ?: throw NotFoundUserException()
```
Comment (uuccu) [inline src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: existingFollow는 아래 follower,followee랑은 다르게 if써서 Exception한 이유가 있음?

File: src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        return userMapStruct.toDtos(followers)
+    }
+
+    @Transactional
+    fun findFolloweesByUserId(
+        userId: Long,
+    ): List<UserDto> {
+        val followees = followRepository.findAllByFollowerId(userId).map { it.followee }
+        return userMapStruct.toDtos(followees)
+    }
```
Comment (uuccu) [inline src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: 조회는 readonly여서 굳이 transactional할 필요가 없음.

File: src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        val followers = followRepository.findAllByFolloweeId(userId).map { it.follower }
+        return userMapStruct.toDtos(followers)
+    }
+
+    @Transactional
+    fun findFolloweesByUserId(
+        userId: Long,
+    ): List<UserDto> {
+        val followees = followRepository.findAllByFollowerId(userId).map { it.followee }
+        return userMapStruct.toDtos(followees)
```
Comment (uuccu) [inline src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: N+1발생할 확률 매우 높으므로 todo n+1 적어쥬삼

**Review by uuccu — APPROVED** [review]
굳 이제 스프링 마스터한듯

**Comment by dicohy27** [comment]
이런 ㅁㅊ 지금확인함ㅋ

File: src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt (hunk 2 above)
Comment (dicohy27) [reply src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: 지워?

Comment (dicohy27) [reply src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: 속도만 느려짐?

File: src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt (hunk 1 above)
Comment (dicohy27) [reply src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: ㄴㄴ없음 무의식인듯 if 지움

File: src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt (hunk 3 above)
Comment (dicohy27) [reply src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt]: ㅇㅇ

**Comment by uuccu** [comment]
### Code Coverage
|File|Coverage [0.00%]|
|:-|:-:|
|[src/main/kotlin/fittibackendapp/domain/follow/entity/Follow.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fentity%2FFollow.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/graphql/mutation/FollowMutationMapper.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fgraphql%2Fmutation%2FFollowMutationMapper.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/graphql/query/FollowQueryMapper.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fgraphql%2Fquery%2FFollowQueryMapper.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fservice%2FFollowService.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/dto/FollowDto.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdto%2FFollowDto.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/exception/AlreadyExistingFollowException.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fexception%2FAlreadyExistingFollowException.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/exception/NotFoundFollowException.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/3b7399149a4af04f4315d2acd7b75e29566936e1/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fexception%2FNotFoundFollowException.kt)|0.00%|

|Total Project Coverage|0.00%|:x:|
|:-|:-:|:-:|

**Comment by uuccu** [comment]
### Code Coverage
|File|Coverage [0.00%]|
|:-|:-:|
|[src/main/kotlin/fittibackendapp/domain/follow/entity/Follow.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fentity%2FFollow.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/graphql/mutation/FollowMutationMapper.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fgraphql%2Fmutation%2FFollowMutationMapper.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/graphql/query/FollowQueryMapper.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fgraphql%2Fquery%2FFollowQueryMapper.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/domain/follow/service/FollowService.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdomain%2Ffollow%2Fservice%2FFollowService.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/dto/FollowDto.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fdto%2FFollowDto.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/exception/AlreadyExistingFollowException.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fexception%2FAlreadyExistingFollowException.kt)|0.00%|
|[src/main/kotlin/fittibackendapp/exception/NotFoundFollowException.kt](https://github.com/FITTI-APP/Fitti-Backend-App/blob/9bf918c77d98fddd228ae9a8dc7eeedd615b98ca/src%2Fmain%2Fkotlin%2Ffittibackendapp%2Fexception%2FNotFoundFollowException.kt)|0.00%|

|Total Project Coverage|0.00%|:x:|
|:-|:-:|:-:|

**Pull request merged**